cdap-data-fabric/src/main/java/co/cask/cdap/data2/nosql/NoSqlStructuredTable.java

```diff
@@ -0,0 +1,310 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.data2.nosql;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.dataset.lib.AbstractCloseableIterator;
+import co.cask.cdap.api.dataset.lib.CloseableIterator;
+import co.cask.cdap.api.dataset.table.Scanner;
+import co.cask.cdap.api.dataset.table.Table;
+import co.cask.cdap.data2.dataset2.lib.table.MDSKey;
+import co.cask.cdap.spi.data.InvalidFieldException;
+import co.cask.cdap.spi.data.StructuredRow;
+import co.cask.cdap.spi.data.StructuredTable;
+import co.cask.cdap.spi.data.table.StructuredTableSchema;
+import co.cask.cdap.spi.data.table.field.Field;
+import co.cask.cdap.spi.data.table.field.FieldFactory;
+import co.cask.cdap.spi.data.table.field.FieldType;
+import co.cask.cdap.spi.data.table.field.Range;
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Nosql structured table implementation.
+ */
+public final class NoSqlStructuredTable implements StructuredTable {
+  private static final Logger LOG = LoggerFactory.getLogger(NoSqlStructuredTable.class);
+  private final Table table;
+  private final StructuredTableSchema schema;
+  private final FieldFactory fieldFactory;
+
+  public NoSqlStructuredTable(Table table, StructuredTableSchema schema) {
```

Comment:
I am a bit worry about implementation of SPI that depends on API (table). Is it the design? Ideally, I think the dependency should be one way like this:

API (interface) -> API implementation which uses SPI -> SPI implementation (based on configuration to pick an implementation).

Seems like the current implementation is doing it in reverse order. Is this correct? @anew

Reply:
After discussing with @chtyim and @anew , we decide to put every implementation of spi in data-fabric to avoid circular dependency. Opened https://issues.cask.co/browse/CDAP-14749 for better hierarchy of the spi.